src/components/KeyCapturer.tsx

```diff
@@ -90,33 +91,59 @@ type KeyCapturerProps = Record<
  * If you want the default behaviour to be preserved (i.e. only want to hook into the event
  * instead of replacing it) set the `propagate${Key}` prop (e.g. propagateBackspace).
  */
-class KeyCapturer extends React.Component<KeyCapturerProps> {
-    _handleKeyEvent = (event: React.KeyboardEvent) => {
+function KeyCapturer(props: KeyCapturerProps) {
+    const { children, eventName = 'onKeyDown' } = props
+    const composingRef = useRef(false)
+    const composingEventHandlers =
+        typeof props.onEnter === 'function'
+            ? {
+                  onCompositionStart: () => {
+                      composingRef.current = true
+                  },
+                  onCompositionEnd: () => {
+                      composingRef.current = false
+                  },
+              }
+            : undefined
+
+    function handleKeyEvent(event: React.KeyboardEvent<HTMLInputElement>) {
         // https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode
         const key =
             event.key !== undefined
                 ? KeyCapturerResolver.resolveByKey(event.key)
                 : KeyCapturerResolver.resolveByKeyCode(event.keyCode)
 
+        if (
+            key === SUPPORTED_KEYS.ENTER &&
+            typeof props.onEnter === 'function'
+        ) {
+            if (
+                composingRef.current ||
+                // Safari fires the onCompositionEnd event before the keydown event, so we
+                // have to rely on the 229 keycode, which is Enter when fired from an IME
+                // https://www.w3.org/TR/uievents/#determine-keydown-keyup-keyCode
+                (event.keyCode || event.which) === 229
+            ) {
+                return
+            }
+        }
+
         if (key && Object.values(SUPPORTED_KEYS).includes(key)) {
-            if (typeof this.props[`on${key}`] === 'function') {
+            if (typeof props[`on${key}`] === 'function') {
                 // @ts-expect-error Dynamic type not expressible in TypeScript.
-                this.props[`on${key}`]()
-                if (this.props[`propagate${key}`] !== true) {
+                props[`on${key}`](event)
+                if (props[`propagate${key}`] !== true) {
                     event.preventDefault()
                     event.stopPropagation()
                 }
             }
         }
     }
 
-    render() {
-        const { children, eventName = 'onKeyDown' } = this.props
-
-        return React.cloneElement(children as React.ReactElement, {
-            [eventName]: this._handleKeyEvent,
-        })
-    }
+    return React.cloneElement(children, {
+        [eventName]: handleKeyEvent,
+        ...composingEventHandlers,
```

Comment:
`composingEventHandlers` could be `undefined`?

Reply:
I think `undefined` can be safely spread into objects (but not into arrays, don't ask me why 🤷‍♂️).

<img width="362" alt="image" src="https://user-images.githubusercontent.com/15199/85899478-24648400-b7cc-11ea-9a09-d535841df27e.png">

Reply:
Yes, it also seems to work for `null` as well as with `Object.assign`, but I can use a fallback empty object for `composingEventHandlers` 👍

Reply:
Ah okay I see. Just learned about it now 👍

https://github.com/tc39/ecma262/issues/687
https://stackoverflow.com/questions/46957194/javascript-es6-spread-operator-on-undefined


> Yes, it also seems to work for null as well as with Object.assign, but I can use a fallback empty object for composingEventHandlers 👍

No need to change the code. I think it is completely fine after knowing it is a defined behavior.

Reply:
Sounds good 😊

I'll wait until it's fully tested on the Twist side before merging and cutting a release here 👍